[Edric Yeo] iP

edricyeo opened the pull request:
# DukePro

> "Your mind is for having ideas, not holding them." - David Allen [(source)](https://dansilvestre.com/productivity-quotes/)

DukePro remembers the things you need to do so you do not have to! It is:
- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is:
1. download it from [here](https://github.com/edricyeo/ip)
2. double-click it
3. add your tasks
4. let it manage your tasks for you! 😍 

And it is **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)
#
If you are a Java programmer, you can use it to practice Java too. Heres the `main` method:
```java
public static void main(String[] args) throws DukeException, IOException {
    String filePath = "data/duke.txt";
    new Duke(filePath).run();
}
```

Commit:
Add Gradle support

edricyeo committed:
first commit

edricyeo committed:
implemented add and list features for level 2

edricyeo committed:
implemented mark as done function for level 3

edricyeo committed:
implemented classes for todos, events, deadlines

edricyeo committed:
implemented testing of text UIs

edricyeo committed:
implemented exception handling (printing of statements)

edricyeo committed:
implemented delete feature, used ArrayList to store tasks instead

edricyeo committed:
managed to get test result to pass after installing dos2unix

edricyeo committed:
Add save feature for level 7

edricyeo committed:
Add date functionality

edricyeo committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

edricyeo committed:
Add more classes for OOP

edricyeo committed:
Update OOP classes

edricyeo committed:
Update parse event method

edricyeo committed:
Add packages for files

edricyeo committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

edricyeo committed:
add gradle files

edricyeo committed:
Add tests for Deadline and Parser class

edricyeo committed:
Create JAR file for duke

edricyeo committed:
Add Javadocs for all public methods

edricyeo committed:
Add find method to look for matching tasks

edricyeo committed:
Fix style

edricyeo committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Storage.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java
#	src/main/java/duke/command/AddCommand.java
#	src/main/java/duke/command/ExitCommand.java
#	src/main/java/duke/command/ListCommand.java
#	src/main/java/duke/command/MarkCommand.java
#	src/main/java/duke/exception/EmptyDateException.java
#	src/main/java/duke/exception/UnknownCommandException.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Task.java

edricyeo committed:
Merge branch 'branch-Level-9'

domoberzin commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,83 @@
+     * @param input The line of text inputted by the user.
+     * @return A Command that performs the task specified by the user.
+     */
+    public static Command parse(String input) throws DukeException {
```
Could add the @throws in the JavaDoc for this method since it throws an exception

domoberzin commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,132 @@
+     *
+     * @return The number of Tasks in the list.
+     */
+    public int size() {
```
Method name should be a verb

domoberzin commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,132 @@
+     * @param idx The index of the task.
+     * @return The Task at the given index.
+     */
+    public Task get(int idx) {
```
Could rename it to getTask to be more specific

domoberzin commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,32 @@
+     * @param ui      The Ui instance for the task manager.
+     * @param storage The Storage instance for the task manager.
+     */
+    public abstract void execute(TaskList tasks, Ui ui, Storage storage) throws DukeException;
```
The abstract class throws an exception, but none of the overriding methods throw the same exception, could possibly remove it since its not utilised.

domoberzin commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,49 @@
+ */
+public class DeleteCommand extends Command {
+
+    private int idx;
```
Could add JavaDoc for this class member to specify that it tracks the index of the task

domoberzin commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,83 @@
+package duke;
+
+import duke.command.*;
```
Import classes should be listed explicitly

domoberzin commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,88 @@
+     *
+     * @return A List of Tasks that have been stored in the data file.
+     */
+    public List<Task> load() throws FileNotFoundException {
```
Could rename it to loadTasks to be specific

domoberzin commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,88 @@
+     *
+     * @param tasks The current List of Tasks.
+     */
+    public void save(TaskList tasks) throws IOException {
```
Could rename it to saveTasks to be specific

domoberzin commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,132 @@
+ */
+public class TaskList {
+
+    private List<Task> tasks;
```
Could add class member JavaDoc

domoberzin reviewed (commented):
More or less good, just some slight nits and class member comments

edricyeo committed:
Fix bugs for find command, fix style and JavaDoc

edricyeo committed:
Fix bugs for date parsing

edricyeo committed:
Fix jar file

wadia82 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,88 @@
+    /**
+     * Method that saves the current tasks in the data file.
+     *
+     * @param tasks The current List of Tasks.
```
Do you think you should add a @throws in JavaDoc as the method throws an exception?

wadia82 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,88 @@
+ */
+public class Storage {
+
+    private String filePath;
```
You could add a JavaDoc for class member. You haven't done that for a few other class members in other classes too.

wadia82 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,139 @@
+ */
+public class Ui {
+
+    private Scanner sc;
```
Do you think it would be better to create the Scanner object in the readCommand method as it is the only function in which it is used?

wadia82 commented on src/main/java/duke/exception/StorageFileNotFoundException.java:
```diff
@@ -0,0 +1,16 @@
+ *
+ * @author Edric Yeo
+ */
+public class StorageFileNotFoundException extends DukeException {
```
You could maybe name the classes that extend from DukeException in the format DukeStorageFileNotFoundException.

wadia82 reviewed (commented):
I think your program is well structured. It is easy to read and understand the flow. I have suggested very minor changes which you can implement if they make sense to you.

edricyeo committed:
Add GUI functionality

edricyeo committed:
Update GUI such that text does not get cut off

edricyeo committed:
Improve GUI

edricyeo committed:
Add assertions

Added assertions to ensure that valid indexes are passed when creating
new MarkCommand, UnmarkCommand, and DeleteCommand instances.

Added assertions to ensure that a valid task from the list is deleted,
marked or unmarked.

Added assertions to ensure that the Ui class has a valid scanner to
read user input

edricyeo committed:
Improve code quality

General improvement in style of code and refactoring of long methods.

Reduced length of code in Parser class by creating methods to process
description and date fields, to make the parse method less lengthy.

Added a new InvalidDateException to handle case where date passed into
a Todo or Deadline is invalid.

edricyeo committed:
Merge pull request #1 from edricyeo/branch-A-CodeQuality

Improve code quality

edricyeo committed:
Merge branch 'branch-A-Assertions'

# Conflicts:
#	src/main/java/duke/Parser.java

edricyeo committed:
Improve code structure for Parser class

edricyeo committed:
Added extension to view tasks in a form of a schedule

edricyeo committed:
Add more exceptions to handle invalid user input

edricyeo committed:
Improve GUI

edricyeo committed:
Add user guide and minor bug fixes

edricyeo committed:
Update storage logic

edricyeo committed:
Improve GUI and fix related bugs

edricyeo committed:
Improve user guide

edricyeo committed:
Update README

edricyeo committed:
Fix JAR file for windows
